components/raftstore/src/store/fsm/peer.rs

```diff
@@ -4417,13 +4417,13 @@ where
 
         let first_idx = self.fsm.peer.get_store().first_index();
 
-        let mut compact_idx = if force_compact
-            // Too many logs between applied index and first index.
-            || (applied_idx > first_idx && applied_idx - first_idx >= self.ctx.cfg.raft_log_gc_count_limit)
-            // Raft log size ecceeds the limit.
+        let mut compact_idx = if force_compact && replicated_idx > first_idx {
+            replicated_idx
+        } else if (applied_idx > first_idx
```

Comment:
When code hits this line, the compact_index will still be larger than replicated_idx. Will it still have problem for RaftEngine?

Reply:
What this PR does is to avoid frequent GC when some TiKVs are down. The GC triggered by size limit will be infrequent.

Reply:
> What this PR does is to avoid frequent GC when some TiKVs are down. The GC triggered by size limit will be infrequent.

Understand. Could you explain why frequent GC triggers the problem, but infrequent GC does not? Also It will be better to add the some brief explanation why the PR is going to fix the problem.

Reply:
Log lag happens when some log entries are not replicated to the follower. If those log entries are compacted in leader, then we are forced to use snapshot to synchronize log, which is costly. When snapshot transfer throughput cannot catch up with incoming foreground writes, log lag can never recover.

This PR makes sure that force compaction can never affect the Raft log catch up behavior, by not deleting any unreplicated entries.

Reply:
For infrequent GC due to its size, could it still delete unreplicated entries and lead to exactly same problem?
I know the probability of it is lower than force compaction, but still possible?

Reply:
That's the same behavior as before. It didn't cause problem then, won't cause problem now. To be more specific, it will only be an issue when snapshot sending rate is slower than GC rate, which isn't likely for size-induced GC.